### PR TITLE
Feature/96 write excel custom style, closes #96, closes #97

### LINKED ISSUE #96: Apply custom styles when writing to Excel workbook
`write_excel()` currently supports a `style` bool parameter, which can be set to `True` to apply a few default, hard-coded font and background colors to table blocks in the output Excel workbook. 

Let's expand this, such that `style` also can accept a detailed style specification. This could be in the form of a JSON-like dict, JSON string, CSS... and/or even a `Table` (the bootstrapping is appealing). 

For example, a rough sketch of a JSON-like dict could be:
```
style={
    "table_name": {
        "font": {"color": "#12f400", "bold": True},   # hex color code
        "background": (212, 212, 212), ...},  # rgb color code
    "destinations": {...}, 
    ...}
```

The `table_name` dict would describe what style to apply to the table name row. (Currently, for example, the default is a blueish bold font with a certain gray background.) Same idea for `destinations`, for the destination row. All in all there would be five top-level keys:

 - `table_name`
 - `destinations`
 - `column_names`
 - `column_units`
 - `values`

I think those five together cover all the parts of a table block. Maybe add a sixth: `num_blank_lines_between_tables` (though find a nicer, shorter name) 

A style specification described thusly would be used on all the table blocks written by a given call of `write_excel()`.

### LINKED ISSUE #97: Center the unit and values "rows" in transposed tables written to Excel
The default style specification in `write_excel()` with `style=True` should cause the horizontal alignment of the text in the unit row and value rows to be set to "center". 

Compare these two rendered tables:
image

The one on the left with default horizontal alignment has values that seem to jump left and right. 

The one on the right with centered units and values is, generally speaking, easier to read. Therefore this should be the default style.

### PR DESCRIPTION
**Pull request by jfcorbett**
Can now optionally specify custom styles in `write_excel()` by passing to parameter `styles` a JSON-like structure of dicts! (#96)

Also, when `write_excel()` parameter `styles` is truthy (i.e. either `True` or a custom dict), transposed tables' column units and values get horizontally centered by default (#97). This default is overridden by any horizontal alignment specified explicitly in `styles`. 

This is all demo'ed in the tests.

Other changes of note:
- `write_excel()` parameter `style` renamed to `styles` because really it's five different styles for each of {tablename, destination, column names, column units, values}, and it makes for a more convenient namespace internally 
- `write_excel()` parameter `num_blank_rows_between_tables` renamed to `sep_lines` for conciseness.

### CODE REVIEW

File: pdtable/io/_excel_openpyxl.py [hunk 1]
```diff
@@ -50,17 +86,19 @@ def write_excel_openpyxl(tables, path, na_rep, style, num_blank_rows_between_tab
         for t in tabs:
             # Keep track of table dimensions for formatting as tuples (num rows, num cols, transposed)
             table_dimensions.append((len(t.df), len(t.df.columns), t.metadata.transposed))
-            _append_table_to_openpyxl_worksheet(t, ws, num_blank_rows_between_tables, na_rep)
+            _append_table_to_openpyxl_worksheet(t, ws, sep_lines, na_rep)
 
-        if style:
-            _format_tables_in_worksheet(ws, table_dimensions, num_blank_rows_between_tables)
+        if styles:
+            styles = DEFAULT_STYLE_SPEC if styles is True else styles
```
Comment (vegatheil) [inline pdtable/io/_excel_openpyxl.py:92]: Should there be a validation of the `styles` dictionary? Or should it simply "not style" if the dictionary is not setup correctly?

File: pdtable/io/_excel_openpyxl.py [hunk 2]
```diff
@@ -79,20 +117,58 @@ def _append_table_to_openpyxl_worksheet(
...
             ws.append(_represent_row_elements(row, units, na_rep))
 
-    for _ in range(num_blank_rows_between_tables):
+    for _ in range(sep_lines):
         ws.append([])  # blank row marking table end
 
 
-def _format_tables_in_worksheet(
-        ws: OpenpyxlWorksheet, table_dimensions: List[Tuple[int, int, bool]], num_blank_rows_between_tables: int
+def deep_get(dictionary, keys, default=None):
```
Comment (vegatheil) [inline pdtable/io/_excel_openpyxl.py:124]: This is obviously implemented close to where it is being used, but would there be a "design" benefit of putting it somewhere more general to be shared across?

File: pdtable/io/_excel_openpyxl.py [hunk 3]
```diff
@@ -79,20 +117,58 @@ def _append_table_to_openpyxl_worksheet(
...
+        deep_get(d, ['meta', 'garbage'], default='-') # => '-'
+    """
+    if dictionary is None:
+        return default
+    if not keys:
+        return dictionary
+    return deep_get(dictionary.get(keys[0]), keys[1:], default)
+
+
+def _style_cells(cells: Iterable[Cell], style: Optional[Dict]) -> None:
```
Comment (vegatheil) [inline pdtable/io/_excel_openpyxl.py:142]: Does the `Optional` type hint automatically include a default value of `None`, e.g. `style: Optional[Dict]=None`?

File: pdtable/io/_excel_openpyxl.py [hunk 4]
```diff
@@ -79,20 +117,58 @@ def _append_table_to_openpyxl_worksheet(
...
+
+    Modified from: https://stackoverflow.com/a/50173148/119775
+
+    Example:
+        d = {'meta': {'status': 'OK', 'status_code': 200}}
+        deep_get(d, ['meta', 'status_code'])          # => 200
+        deep_get(d, ['garbage', 'status_code'])       # => None
+        deep_get(d, ['meta', 'garbage'], default='-') # => '-'
+    """
+    if dictionary is None:
```
Comment (vegatheil) [inline pdtable/io/_excel_openpyxl.py:135]: Does it make sense to add tests of these first two `if`s, even though these are of course inherently part of the existing tests, though only because it is recursive?

**Review by vegatheil — CHANGES_REQUESTED** [review]

File: pdtable/io/_excel_openpyxl.py (hunk 1 above)
Comment (jfcorbett) [reply pdtable/io/_excel_openpyxl.py:92]: Hm, good question. Right now it looks for styles in the dict; and applies them if it finds them. If there's a typo in the style keys (e.g. `fnot` instead of `font`, or `taaaable_name` instead of `table_name`), then that part just gets ignored, just as irrelevant dictionary keys are (e.g. an `'hello_world'` entry can be there, but it gets ignored). 

If, however, the _value_ is bad (e.g. `'color': 'my_favorite_color  # <<< not a valid RGB hex code!' or `'alignment': {'horizontal': 'bogus'}  # <<< unknown horizontal alignment!`), then this will fail hard. We could add some validation at the input, but covering all possible valid font, fill, alignment, etc. property values would be a daunting amount of work. 

I think the middle ground here may be to do better error handling. E.g. try applying horizontal alignment, catch any obscure error thrown by openpyxl, and raise our own, more intelligible error. I'll do that. 

If

Comment (jfcorbett) [reply pdtable/io/_excel_openpyxl.py:92]: Done. Now the user will get a friendly error message from which they can infer where their style spec has gone awry. Will push soon.

File: pdtable/io/_excel_openpyxl.py (hunk 2 above)
Comment (jfcorbett) [reply pdtable/io/_excel_openpyxl.py:124]: If it's going to be reused elsewhere, then absolutely yes. Until then, I'm leaning toward leaving it be. If we move it somewhere more general, then it's going to be a module at the top level, with a single function in it... not worth the top-level clutter at this stage, IMHO.

File: pdtable/io/_excel_openpyxl.py (hunk 4 above)
Comment (jfcorbett) [reply pdtable/io/_excel_openpyxl.py:135]: IMHO, not really. We should test the behaviour that we actually need. As consumer of this function, do I expect anything specific to happen if I pass it a `dictionary=None` or empty `keys`? Not really, because I don't see any scenario in which I would want to do that.

File: pdtable/io/_excel_openpyxl.py (hunk 3 above)
Comment (jfcorbett) [reply pdtable/io/_excel_openpyxl.py:142]: No, it doesn't. Type hints don't get compiled and as such they have no bearing on default values.

File: pdtable/io/_excel_openpyxl.py [hunk 5]
```diff
@@ -102,27 +178,41 @@ def _format_tables_in_worksheet(
...
-            name_cells = table_rows[2]
-            unit_cells = table_rows[3]
-
-        header_row = table_rows[0]
-        destination_row = table_rows[1]
-        _format_cells(header_row, font=header_font, fill=header_fill)
-        _format_cells(destination_row, font=destination_font, fill=header_fill)
-        _format_cells(name_cells, font=name_font, fill=variable_fill)
-        _format_cells(unit_cells, fill=variable_fill)
+            column_name_cells = table_rows[2]
```
[the remark this answers is not in the document]
Comment (jfcorbett) [reply pdtable/io/_excel_openpyxl.py:196]: Yeah, this was bothering me as well. `table_name` and `column_name` need to be differentiated, so the `column_` prefix is necessary there, but it's true that for consistency, it should either be (`column_units`, `column_values`) (consistent style but verbose) or (`units`, `values`) (not quite as consistent, but succinct), not a mix of the two like I did. 

Right now I'm leaning toward brevity i.e. (`units`, `values`). I'll convert to that. 

I think I concur with you that

File: pdtable/io/_excel_openpyxl.py (hunk 3 above)
Comment (vegatheil) [reply pdtable/io/_excel_openpyxl.py:142]: Let me rephrase: can the method be called as `_style_cells(cells=my_cells)` or will it fail from not having explicitly passed a `None`-`style` parameter?

**Comment by jfcorbett** [comment]
@vegatheil I think I addressed most of your requested changes. How do we feel about this PR now?

File: pdtable/io/_excel_openpyxl.py (hunk 3 above)
Comment (jfcorbett) [reply pdtable/io/_excel_openpyxl.py:117]: It will fail.
This is a single-use, throwaway method, so I didn't worry too much about the ergononmics of various use cases :-)

File: pdtable/io/_excel_openpyxl.py [hunk 6]
```diff
@@ -9,14 +10,24 @@
...
 from openpyxl.utils import get_column_letter
 
 from pdtable import Table
 from pdtable.io._represent import _represent_row_elements, _represent_col_elements
 
-
 DEFAULT_SHEET_NAME = "Sheet1"
+DEFAULT_STYLE_SPEC = {
+    "table_name": {
+        "font": {"color": "1F4E78", "bold": True,},  # hex color code
```
Comment (vegatheil) [inline pdtable/io/_excel_openpyxl.py:23]: RGB or hex color code? Or is it the same?

**Review by vegatheil — APPROVED** [review]

File: pdtable/io/_excel_openpyxl.py (hunk 6 above)
Comment (jfcorbett) [reply pdtable/io/_excel_openpyxl.py:23]: I guess it's both... RRGGBB where RR, GG, and BB are each hex numbers between 00 and FF
Not sure what the usual terminology is, but I've called them RGB hex codes :-)
